swift/common/middleware/ephemeral_object.py

```diff
@@ -0,0 +1,385 @@
+# Copyright (c) 2010-2020 OpenStack Foundation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import time
+
+from paste.deploy import loadwsgi
+from six.moves.urllib.parse import parse_qs
+from swift.common.swob import Request, HTTPInternalServerError, HTTPBadRequest
+from swift.common.utils import get_logger, MD5_OF_EMPTY_STRING, \
+    split_path, parse_connection_string
+from swift.common.wsgi import loadcontext, PipelineWrapper
+
+from oio.common.redis_conn import catch_service_errors
+from oio.common.redis_conn import RedisConnection
+
+LOG = None
+MIDDLEWARE_NAME = 'ephemeral-object'
+DEFAULT_LIMIT = 10000
+
+DELIMITER = '/'
+SWIFT_SOURCE = 'EO'
+
+# For each UploadID, two redis keys will be created:
+# - one entry in zset MPU_LIST:{ACCOUNT}:{CONTAINER/BUCKET}: it will be used
+# to check if UploadId is still present and to support list-multipart-uploads
+# - one key as MPU_HDRS:{ACCOUNT}:{CONTAINER/BUCKET}:{UPLOADID} to hold headers
+LST_PREFIX = 'EO_LIST:'
+HDR_PREFIX = 'EO_HDRS:'
+
+# When iterating in Redis with non recursive listing, we force next
+# zrangebylex to skip already used prefix
+FAKE_MARKER = "\xef\xbf\xbd"
```

Comment:
In python3, I think it should be `\ufffd`.

Reply:
`u"\ufffd"` works for both Python 2 and 3.